Objects/rangeobject.c

```diff
@@ -979,75 +998,89 @@ typedef struct {
 static PyObject *
 longrangeiter_len(longrangeiterobject *r, PyObject *no_args)
 {
-    Py_INCREF(r->len);
-    return r->len;
+    PyObject *len;
+    Py_BEGIN_CRITICAL_SECTION(r);
+    len = Py_NewRef(r->len);
+    Py_END_CRITICAL_SECTION();
+    return len;
 }
 
 static PyObject *
 longrangeiter_reduce(longrangeiterobject *r, PyObject *Py_UNUSED(ignored))
 {
     PyObject *product, *stop=NULL;
-    PyObject *range;
+    PyObject *range, *result=NULL;
 
+    Py_BEGIN_CRITICAL_SECTION(r);
     /* create a range object for pickling.  Must calculate the "stop" value */
     product = PyNumber_Multiply(r->len, r->step);
     if (product == NULL)
-        return NULL;
+        goto fail;
     stop = PyNumber_Add(r->start, product);
     Py_DECREF(product);
     if (stop ==  NULL)
-        return NULL;
+        goto fail;
     range =  (PyObject*)make_range_object(&PyRange_Type,
                                Py_NewRef(r->start), stop, Py_NewRef(r->step));
     if (range == NULL) {
         Py_DECREF(r->start);
         Py_DECREF(stop);
         Py_DECREF(r->step);
-        return NULL;
+        goto fail;
     }
 
     /* return the result */
-    return Py_BuildValue("N(N)O", _PyEval_GetBuiltin(&_Py_ID(iter)),
-                         range, Py_None);
+    result = Py_BuildValue("N(N)O", _PyEval_GetBuiltin(&_Py_ID(iter)),
+                           range, Py_None);
+fail:
+    ; // A statement must follow the label before Py_END_CRITICAL_SECTION.
+    Py_END_CRITICAL_SECTION();
+    return result;
 }
 
 static PyObject *
 longrangeiter_setstate(longrangeiterobject *r, PyObject *state)
```

Comment:
If we need to wrap functions in critical sections, I think we should generally convert them to argument clinic, if possible, and use the `@critical_section` annotation to keep the boilerplate mostly in a generated file.

That's not yet possible for `longrangeiter_next` (because it's a tp_iternext, not a PyMethodDef entry), but it's doable for the other functions.

Reply:
I've dropped the range iterator fixes (fast and long) from this PR, and I'll leave it for a separate change. (I'll file an issue.)